### PR TITLE
Functionality/pickling for commhooks

### PR DESCRIPTION
**Pull request by aovladi**
This PR addresses issue address #75666.  
Stateful communication hook now can be saved and reloaded to resume training.

Current PR adds the functionality for PowerSGD communication hook and tests that communication hook can be properly saved and restored.

PowerSGD implementation uses ``__slots__``, as a result introduced __getstate__ and __setstate__ methods are implemented to work with `__slots__` and not` __dict__`.

`__getstate__ `
	
	 Returns:
           A dictionary that represents a ``PowerSGDState`` which will be pickled and saved.
          ``process_group`` is non-serializable and excluded from a returned state.

`__setstate__`

	Takes a provided ``state`` and retrieves ``PowerSGDState``.
        ``process_group`` is set to default with a proper warning issued to a user.

Unit test

A hook-independent `_test_hook_pickling` is added with this PR, as well as `test_ddp_hook_pickling_powerSGD`, which tests `powerSGD`’s ability to be saved and reloaded.

Currently, the test creates a ddp model with a provided hook, trains it for 10 epochs and saves model’s state and hook’s state. 
During reloading, unit test makes sure that a warning was logged (only one warning and the proper one). It then proceeds to check that reloaded hook and original hook are the same. Finally, it checks that a hook’s state was properly initialized:
	- it compares slot values (all, but 2: `process_group` and `rng`) for original and reloaded state
	- it checks that process group was set to a default group
	- it checks that a random state was restored properly with np.testing.assert_array_equal, because `rng` is an instance of `np.random.RandomState`, represented by a tuple. One of entries is of `ndarray dtype[uint32]` type and `np.testing.assert_array_equal` is used for assertion. 

Future To-Do:
	- Implement similar __getstate__ and __setstate__ for other stateful communication hooks
	- Add appropriate tests

### CODE REVIEW

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 1]
```diff
@@ -264,6 +266,31 @@ def __init__(
         # Turn on if compression / decompression computation is a bottleneck.
         self.batch_tensors_with_same_shape = batch_tensors_with_same_shape
 
+    def __getstate__(self):
+        r"""
+        Returns a ``PowerSGDState`` which will be pickled and saved.
```
Comment (awgu) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:271]: nit: I think the return type is not `PowerSGDState` but rather `Dict[str, Any]`?

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 2]
```diff
@@ -264,6 +266,31 @@ def __init__(
...
+        ``process_group`` is not serializable and excluded from
+        a returned state.
+        """
+        slots = copy.copy(self.__slots__)
+        slots.remove("process_group")
+        return {
+            slot: getattr(self, slot)
+            for slot in slots if hasattr(self, slot)
+        }
```
Comment (awgu) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:280]: Could we do:
```suggestion
        return {
            slot: getattr(self, slot)
            for slot in slots if hasattr(self, slot) and slot != "process_group"
        }
```
I think you are creating a new `list` with `slots = copy.copy(self.__slots__)`, but we can avoid this shallow copy and iterate directly over the existing `self.__slots__`.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 3]
```diff
@@ -1,13 +1,15 @@
 from collections import defaultdict
 import logging
 import math
+import copy
```
Comment (awgu) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:4]: nit: sort imports (there should be a menu option in VS Code)

File: torch/testing/_internal/distributed/distributed_test.py [hunk 4]
```diff
@@ -6,6 +6,7 @@
 import sys
 import tempfile
 import time
+import pickle
```
Comment (awgu) [inline torch/testing/_internal/distributed/distributed_test.py:9]: nit: sort

File: torch/testing/_internal/distributed/distributed_test.py [hunk 5]
```diff
@@ -8820,4 +8821,108 @@ def celu_hook(module, _input, output):
...
             self.assertEqual(ddp_grads[1], local_model.fc.bias.grad)
 
+        def _test_hook_pickling(self, hook, hook_state):
+            learning_rate = 0.01
+            chkpt_file = tempfile.gettempdir() + "/checkpoint.pt"
+            rank = self.rank
+
+            input = torch.randn(1, 1, device=rank)
+            net = torch.nn.Linear(1, 5).to(rank)
+            ddp_model = torch.nn.parallel.DistributedDataParallel(
```
Comment (awgu) [inline torch/testing/_internal/distributed/distributed_test.py:8831]: nit: 
```suggestion
            ddp_model = DistributedDataParallel(
```
(https://github.com/pytorch/pytorch/blob/8cea5e63def0352a1e72455ec87fec40d3cbab8a/torch/testing/_internal/distributed/distributed_test.py#L47)

File: torch/testing/_internal/distributed/distributed_test.py [hunk 6]
```diff
@@ -8820,4 +8821,108 @@ def celu_hook(module, _input, output):
...
+            for _ in range(10):
+                optimizer.zero_grad()
+                loss = ddp_model(input).sum()
+                loss.backward()
+                optimizer.step()
+
+            state = {
+                'state_dict': ddp_model.state_dict(),
+                'comm_hook': pickle.dumps(hook),
+                'comm_hook_state': pickle.dumps(hook_state)
```
Comment (awgu) [inline torch/testing/_internal/distributed/distributed_test.py:8851]: I am curious if we can do
```suggestion
                'comm_hook': hook,
                'comm_hook_state': hook_state
```
and then `torch.save()` will internally recurse to pickle the `"comm_hook"` and `"comm_hook_state"`. In other words, I wonder if it is necessary to manually call `pickle.dumps()`.

(I mention this because this will be the expected API usage from the user, and preferably, the user does not need to manually call `pickle.dumps()`.)

File: torch/testing/_internal/distributed/distributed_test.py [hunk 7]
```diff
@@ -8820,4 +8821,108 @@ def celu_hook(module, _input, output):
...
+
+            hook = powerSGD.powerSGD_hook
+            powersgd_state = powerSGD.PowerSGDState(
+                process_group=None,
+                matrix_approximation_rank=1,
+                start_powerSGD_iter=4,
+            )
+            self._test_hook_pickling(hook, powersgd_state)
+
+
```
Comment (awgu) [inline torch/testing/_internal/distributed/distributed_test.py:8923]: nit: I think we only need to leave two newlines between the end of he test and the `instantiate_parametrized_tests()`,

File: torch/testing/_internal/distributed/distributed_test.py [hunk 8]
```diff
@@ -8820,4 +8821,108 @@ def celu_hook(module, _input, output):
...
+                "NOTE: PowerSGDState is set to use the default process group."
+            )
+            # Check that loaded function is correct
+            self.assertEqual(dummy_hook.__qualname__, hook.__qualname__)
+
+            slots_original = hook_state.__slots__
+            slots_restored = dummy_hook_state.__slots__
+            self.assertEqual(slots_original, slots_restored)
+
+            slots = copy.copy(slots_restored)
```
Comment (awgu) [inline torch/testing/_internal/distributed/distributed_test.py:8881]: nit: Any reason to shallow copy the `list` and remove entries instead of using `if entry != "process_group" and entry != "rng"` when iterating over `slots_restored`?

File: torch/testing/_internal/distributed/distributed_test.py [hunk 9]
```diff
@@ -8820,4 +8821,108 @@ def celu_hook(module, _input, output):
...
+
+            if rank == 0:
+                torch.save(state, chkpt_file)
+
+            dist.barrier()
+            map_location = {'cuda:%d' % 0: 'cuda:%d' % rank}
+            checkpoint = torch.load(chkpt_file, map_location=map_location)
+
+            dummy_ddp_model.load_state_dict(checkpoint['state_dict'])
+            dummy_hook = pickle.loads(checkpoint['comm_hook'])
```
Comment (awgu) [inline torch/testing/_internal/distributed/distributed_test.py:8862]: Similar to the comment with saving. I wonder if we can just do `checkpoint = torch.load(...)` and set `dummy_hook = checkpoint['comm_hook']` here (and similar for `dummy_hook_state`).

**Review by awgu — COMMENTED** [review]
The PR looks good. I left some minor nits and two actual feedbacks:
- Investigate if we have to manually call `pickle.dumps()`/`pickle.loads()`. `torch.save()`/`torch.load()` call into pickle save and load subroutines (though with an older default protocol level but that should not matter and I am mention that as an fyi).
- Should we consider adding a unit test where the hook state does not have all `__slots__` entries? I noticed that in `__getstate__()`, you have an `if hasattr(self, slot)` check. I am curious if this will work if some `slot` attributes are missing for both save and load.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py (hunk 1 above)
Comment (aovladi) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:271]: This is a valid point. Thanks for correcting. I've changed it in the updated PR.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py (hunk 2 above)
Comment (aovladi) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:280]: I agree, removed unnecessary copy from the updated version.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py (hunk 3 above)
Comment (aovladi) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:4]: Thanks for pointing out this option in VS Code. Sorted inputs in the updated PR.

File: torch/testing/_internal/distributed/distributed_test.py (hunk 4 above)
Comment (aovladi) [reply torch/testing/_internal/distributed/distributed_test.py:9]: I've removed this import, in the updated version it is unnecessary

File: torch/testing/_internal/distributed/distributed_test.py (hunk 5 above)
Comment (aovladi) [reply torch/testing/_internal/distributed/distributed_test.py:8831]: Good point! I've updated PR according to suggestion.

File: torch/testing/_internal/distributed/distributed_test.py (hunk 6 above)
Comment (aovladi) [reply torch/testing/_internal/distributed/distributed_test.py:8851]: I appreciate you sharing an insight about `torch.save()/torch.load()`. I've looked into implementation and checked with unit test, and confirmed, that we don't need to explicitly call `pickle.dumps()` and `pickle.loads()`. I've updated PR accordingly.

File: torch/testing/_internal/distributed/distributed_test.py (hunk 7 above)
Comment (aovladi) [reply torch/testing/_internal/distributed/distributed_test.py:8923]: Noted, thanks!

File: torch/testing/_internal/distributed/distributed_test.py (hunk 9 above)
Comment (aovladi) [reply torch/testing/_internal/distributed/distributed_test.py:8862]: Updated. Left a comment regarding this earlier.

File: torch/testing/_internal/distributed/distributed_test.py (hunk 8 above)
Comment (aovladi) [reply torch/testing/_internal/distributed/distributed_test.py:8881]: I agree that shallow copy is unnecessary. Updated according to suggestions.

**Comment by aovladi** [comment]
@awgu Thank you for comments!
Regarding the first proposal to investigate if we need to call `pickle.dumps(...)` explicitly:  We don't and thanks for pointing this out. I've updated PR based accordingly.

As to the `hasattr` and appropriate unittest. We've discussed this offline: `hasattr` is actually unnecessary, because we are iterating over `self.__slots__`. As for the unit test, when a corrupted state is loaded, to my understanding and some tests I ran, it is not trivial to check which attributes provided state is missing (unless we know for sure, e.g. `process_group`). My best judgement tells me, that if user wants to save a state, he won't access `__slots__` and remove any attributes. However, please, let me know if I am wrong and I need to dive deeper and find a solution.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 10]
```diff
@@ -264,6 +265,29 @@ def __init__(
...
+        }
+
+    def __setstate__(self, state):
+        r"""
+        Takes a provided ``state`` and retrieves ``PowerSGDState``.
+        ``process_group`` is set to default.
+        """
+        self.process_group = distributed_c10d._get_default_group()
+        logger.warning(
+            "NOTE: Current process group is set to default."
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:286]: nit: Change the warning to something like

`Process group will be set to default group (whole world) when loading PowerSGD state`. If a different group is desired, please set `self.process_group` after PowerSGD state is loaded`.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py [hunk 11]
```diff
@@ -264,6 +265,29 @@ def __init__(
         # Turn on if compression / decompression computation is a bottleneck.
         self.batch_tensors_with_same_shape = batch_tensors_with_same_shape
 
+    def __getstate__(self):
+        r"""
+        Returns a ``Dict[str, Any]`` which will be pickled and saved.
+        ``process_group`` is not serializable and excluded from
```
Comment (rohan-varma) [inline torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:271]: Add a logger warning here indicating this?

File: torch/testing/_internal/distributed/distributed_test.py [hunk 12]
```diff
@@ -8820,4 +8820,100 @@ def celu_hook(module, _input, output):
...
+
+            # Check that the logger has only one entry
+            self.assertEqual(len(captured.records), 1)
+            # Check that the logger has an expected entry
+            self.assertEqual(
+                captured.records[0].getMessage(),
+                'NOTE: Current process group is set to default.'
+            )
+
+            dummy_ddp_model.load_state_dict(checkpoint['state_dict'])
```
Comment (rohan-varma) [inline torch/testing/_internal/distributed/distributed_test.py:8869]: It might be valuable to register the loaded hook + state here and ensure that works without any issues.

Further, the test could be more comprehensive if we run some iterations on this reloaded version and ensure we have parity with before checkpoint (you can check the gradients, and make sure everything starts off / is seeded the same)?

**Review by rohan-varma — COMMENTED** [review]
Awesome work! LGTM overall, a couple of suggestions on the test, and please fix lint issues.

Let's also enhance the documentation: https://github.com/pytorch/pytorch/blob/master/docs/source/ddp_comm_hooks.rst to indicate how to checkpoint comm hooks as part of this task:

- Add a section at the bottom "Checkpointing of comm hooks" similar to "Debugging comm hooks"
- Describe how this should be done (stateful hooks need pickling methods implemented)
- Mention this is done for PowerSGD and can be used as a reference
- Add an end to end example (similar to the test where we save / load a comm hook)

**Comment by rohan-varma** [comment]
Also, can we identify the rest of the stateful comm hooks we offer in-tree and file a single issue mentioning to implement __setstate__ and __getstate__ for them? Would make for great ramp up tasks down the road.

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py (hunk 10 above)
Comment (aovladi) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py]: Warning message is updated

File: torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py (hunk 11 above)
Comment (aovladi) [reply torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:271]: Added logger warning and a check in the unit test.

File: torch/testing/_internal/distributed/distributed_test.py (hunk 12 above)
Comment (aovladi) [reply torch/testing/_internal/distributed/distributed_test.py:8884]: Thanks for suggestion! I've refactored unit test -- i.e. adjusted model initialization and training procedure, added extra training after a model was reloaded, added check on gradients as suggested.

**Comment by aovladi** [comment]
The last commit adds enhanced ddp_comm_hooks.rst as per @rohan-varma suggestion.  A new section "Checkpointing of comm hook" is added towards the end of a file. It describes what methods need to be implemented for proper serialization, as well as warns, that non-serializable attributes should be excluded/properly restored. Finally, an end to end example of saving and reloading is provided in the end of the section.

**Review by rohan-varma — COMMENTED** [review]
awesome work! LGTM overall, just looks like there are doc CI errors, can you fix those? 

```
/opt/conda/lib/python3.7/site-packages/torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:docstring of torch.distributed.algorithms.ddp_comm_hooks.powerSGD_hook.PowerSGDState.__getstate__:1: WARNING: duplicate object description of torch.distributed.algorithms.ddp_comm_hooks.powerSGD_hook.PowerSGDState.__getstate__, other instance in ddp_comm_hooks, use :noindex: for one of them
/var/lib/jenkins/workspace/docs/source/ddp_comm_hooks.rst:160: WARNING: Definition list ends without a blank line; unexpected unindent.
/var/lib/jenkins/workspace/docs/source/ddp_comm_hooks.rst:171: WARNING: Block quote ends without a blank line; unexpected unindent.
/var/lib/jenkins/workspace/docs/source/ddp_comm_hooks.rst:177: WARNING: Block quote ends without a blank line; unexpected unindent.
/var/lib/jenkins/workspace/docs/source/ddp_comm_hooks.rst:186: WARNING: Definition list ends without a blank line; unexpected unindent.
```

**Review by rohan-varma — APPROVED** [review]
LGTM, thanks for adding this!

**Review by awgu — APPROVED** [review]
PR looks good to me! As discussed offline, just double check that the docs render correctly in CI as well in case there are versioning differences between your Mac and CI.

File: torch/testing/_internal/distributed/distributed_test.py [hunk 13]
```diff
@@ -8821,4 +8821,136 @@ def celu_hook(module, _input, output):
...
+            # Excluding ``process_group`` and ``rng``.
+            for entry in dummy_hook_state.__slots__:
+                if entry != "process_group" and entry != "rng":
+                    self.assertEqual(getattr(dummy_hook_state, entry), getattr(hook_state, entry))
+
+            # Check that ``process_group`` was set to default
+            self.assertEqual(dummy_hook_state.process_group, _get_default_group())
+
+            # Check that a random state was restored properly:
+            # :met:`np.random.RandomState.get_state` returns a tuple with entries:
```
[the remark this answers is not in the document]
Comment (aovladi) [reply torch/testing/_internal/distributed/distributed_test.py:8905]: That would be `meth` , I've removed this from comments.
